src/OrchardCore/OrchardCore.Data/SessionWrapper.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Threading.Tasks;
+using OrchardCore.Environment.Shell.Scope;
+using YesSql;
+using YesSql.Indexes;
+
+namespace OrchardCore.Data
+{
+    public class SessionWrapper : ISession
+    {
+        private readonly ISession _session;
+
+        private readonly HashSet<object> _saved = new HashSet<object>();
+
+        private bool _canceled;
+
+        public SessionWrapper(ISession session)
+        {
+            _session = session;
+
+            ShellScope.RegisterBeforeDispose(scope => BeforeDisposeAsync());
+        }
+
+        public Task BeforeDisposeAsync()
+        {
+            if (_session == null)
+            {
+                return Task.CompletedTask;
+            }
+
+            if (!_canceled)
+            {
+                // A save followed by a query, and then a mutation, needs another save to persist the last mutation.
+                foreach (var item in _saved)
```

Comment:
This is doing an extra save no matter what. Does it actually issues an extra Update in the db, or simply a check that the object hasn't changed (yessql does that I think).

Reply:
As i remember, if no query was done it is a no-op, otherwise i think it does an extra update

> or simply a check that the object hasn't changed (yessql does that I think).

I didn't see that.

Then i think it is the same if we save an item1, then save an item2 and do an explicit `commtAsync()` (as in UserStore), here there will be extra updates of item1 and item2, only updates not creations as `_identityMap` is only cleared on `ReleaseSession()` when disposing the session.